### ISSUE TITLE
check_gradient always fails

### ISSUE THREAD

**Issue by fph** [JuliaManifolds/Manopt.jl#245]
Is `check_gradient` working properly? I don't seem to have it return true, even on simple functions.

Even taking the sample `f` and `grad_f` from the [docs](https://manoptjl.org/stable/tutorials/Optimize!/) fails:
```julia
using Manopt, Manifolds, Random, LinearAlgebra
Random.seed!(42);
n = 100
σ = π / 8
M = Sphere(2)
p = 1 / sqrt(2) * [1.0, 0.0, 1.0]
data = [exp(M, p,  σ * rand(M; vector_at=p)) for i in 1:n];
f(M, p) = sum(1 / (2 * n) * distance.(Ref(M), Ref(p), data) .^ 2)
grad_f(M, p) = sum(1 / n * grad_distance.(Ref(M), data, Ref(p)));
check_gradient(M, f, grad_f, throw_error=true)
```
returns
```julia
ERROR: DomainError with 2.687604283142213e-16:
The vector [-1.001187263626093, 0.1028991640105687, -0.887807346557081] is not a tangent vector to [-0.42344094610973243, 0.7120699114881195, 0.5600483964007007] on Sphere(2, ℝ), since it is not orthogonal in the embedding.
Stacktrace:
 [1] is_vector(::ManifoldsBase.TraitList{IsEmbeddedManifold, ManifoldsBase.TraitList{IsDefaultMetric{EuclideanMetric}, ManifoldsBase.TraitList{IsMetricManifold, ManifoldsBase.EmptyTrait}}}, M::Sphere{2, ℝ}, p::Vector{Float64}, X::Vector{Float64}, te::Bool, cbp::Bool; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ ManifoldsBase ~/.julia/packages/ManifoldsBase/ROusQ/src/decorator_trait.jl:481
 [2] is_vector
   @ ~/.julia/packages/ManifoldsBase/ROusQ/src/decorator_trait.jl:435 [inlined]
 [3] is_vector(_t::ManifoldsBase.TraitList{IsIsometricEmbeddedManifold, ManifoldsBase.TraitList{IsEmbeddedManifold, ManifoldsBase.TraitList{IsDefaultMetric{EuclideanMetric}, ManifoldsBase.TraitList{IsMetricManifold, ManifoldsBase.EmptyTrait}}}}, M::Sphere{2, ℝ}, p::Vector{Float64}, X::Vector{Float64}, te::Bool, cbp::Bool; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ ManifoldsBase ~/.julia/packages/ManifoldsBase/ROusQ/src/nested_trait.jl:313
 [4] is_vector(_t::ManifoldsBase.TraitList{IsIsometricEmbeddedManifold, ManifoldsBase.TraitList{IsEmbeddedManifold, ManifoldsBase.TraitList{IsDefaultMetric{EuclideanMetric}, ManifoldsBase.TraitList{IsMetricManifold, ManifoldsBase.EmptyTrait}}}}, M::Sphere{2, ℝ}, p::Vector{Float64}, X::Vector{Float64}, te::Bool, cbp::Bool)
   @ ManifoldsBase ~/.julia/packages/ManifoldsBase/ROusQ/src/nested_trait.jl:308
 [5] is_vector(M::Sphere{2, ℝ}, p::Vector{Float64}, X::Vector{Float64}, te::Bool, cbp::Bool; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ ManifoldsBase ~/.julia/packages/ManifoldsBase/ROusQ/src/nested_trait.jl:306
 [6] is_vector (repeats 2 times)
   @ ~/.julia/packages/ManifoldsBase/ROusQ/src/nested_trait.jl:305 [inlined]
 [7] check_gradient(M::Sphere{2, ℝ}, f::Function, grad_f::Function, p::Vector{Float64}, X::Vector{Float64}; gradient::Vector{Float64}, check_vector::Bool, throw_error::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ Manopt ~/.julia/packages/Manopt/Wl0jp/src/helpers/checks.jl:187
 [8] top-level scope
   @ REPL[37]:1
```
Apparently the problem is that `2.687604283142213e-16` is not exactly zero?

**Comment by kellertuer** [comment]
Ah that looks like your gradient is indeed correct, yes.

The thing that is inprecise here is the tangent vector representing the gradient. Since that is a bit hard to set a tolerance for that you could set

 We have a strict default, just in case, but you can set

```
julia> check_gradient(M, f, grad_f; throw_error=true, check_vector=false)
true
```

and it confirms that your gradient is correct.

**Comment by kellertuer** [comment]
Does the answer resolve your problem? One could – alternatively – introduce a keyword for the vector tolerances (maybe just absolute but maybe also relative – or pass any kwargs down – I‘d have o check); but I feel that the function already has quite a few keywords, that even I tend to forget.

**Comment by mateuszbaran** [comment]
I think `check_vector` should just have a non-zero absolute tolerance, that would likely solve this problem. Anyway, checking the vector is not a key part of this function so maybe even `check_vector` could be `false` by default?

**Comment by kellertuer** [comment]
For now I could also introduce a tolerance here for the vector checks, sure; In general that should be done over in Manifolds.jl.

And sure setting that to false by default would also be possible. I will give that a thought; since that loosens the requirements, it can also be considered non-breaking.

**Comment by kellertuer** [comment]
I opted for the assumption that `is_vector` accepts kwargs always and is able to accept hence `atol` and `rtol` to pass that down to its inner approx.

In general I noticed that for example the sphere one currently passes _all_ kwargs down to atol, that is maybe a bit dangerous. But in a more safer way (`is-vector` accepts all kwargs and uses the ones it needs) it might be nicer even.

But however. The PR I just opened allows to resolve this (keeping `check_vector=true` by default) by setting e.g. `atol=1e-15` or an `rtol` in the `check_gradient`, since it now accepts those two for these checks.
